### ISSUE TITLE
Best way to inject custom request handling logic?

### ISSUE THREAD

**Issue by Tadimsky** [ThreeMammals/Ocelot#897]
We're looking at using Ocelot as a frontend to our services, but one of the things we'd like to be able to do is to take take over the request at some point if it matches certain criteria.

We need to do this to support data residency for our services. If a user's data location is set as some other country, then we need to be able to proxy the request to that country instead. 
We need the user to have authenticated first (so that we can lookup their location), so it would be nice to inject this handler somewhere in the Ocelot pipeline, but it doesn't seem like that's something that could be done right now from what I can see.

We could do middleware before Ocelot even runs (and only use Ocelot for users that have their data stored in the country that the cluster is running in), but I'd love to integrate it more with Ocelot if possible.

I guess the other way of doing this would be to have a custom implementation of the part that gets the downstream route, but that also looks like it's too early in the pipe.

https://github.com/ThreeMammals/Ocelot/blob/41d4f9c6dfa3069dc9035e2e82d0c46f7f08cf56/src/Ocelot/Middleware/Pipeline/OcelotPipelineExtensions.cs#L50

Or potentially of `DownstreamUrlCreatorMiddleware` .

Would something like this be possible? How would you folks recommend doing this kind of thing?

**Comment by yuft** [comment]
Seems like you need a proxy, maybe take a look at https://github.com/damianh/ProxyKit instead.

**Comment by Tadimsky** [comment]
@yuft Yeah, that's what we're using right now actually and it's pretty good. Just a little more entangled with our services than we want it to be.
But yeah, definitely seems like Ocelot would be better as something to sit behind our Proxykit reverse proxy for traffic into the cluster.
